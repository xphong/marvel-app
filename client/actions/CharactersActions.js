import axios from 'axios';

import * as types from '../constants/ActionTypes';
import { ENDPOINT, CHARACTER_LIMIT } from '../constants/AppConstants';

function requestCharactersByName() {
  return {
    type: types.REQUEST_CHARACTERS_BY_NAME
  };
}

function receiveCharacters(data) {
  return{
    type: types.RECEIVE_CHARACTERS,
    payload: {
      data
    }
  };
}

function receiveError(data) {
  return {
    type: types.RECEIVE_CHARACTERS_ERROR,
    payload: {
      data
    }
  };
}

function createCharacterData(data) {
  let characterData = {};

  characterData.name = data.name;
  characterData.url = (data.urls[1] ? data.urls[1].url : data.urls[0].url);
  characterData.image = data.thumbnail.path + '.' + data.thumbnail.extension;
  characterData.description =  (data.description === '' ? 'No description listed for this character.' : characterData.description = data.description);

  return characterData;
}

export function fetchCharactersByName(name) {
  const url = `${ENDPOINT}characters?limit=${CHARACTER_LIMIT}&name=${name}`;
  const opts = {
    url: url,
    timeout: 10000,
    method: 'get',
    responseType: 'json'
  };

  return dispatch => {
    dispatch(requestCharactersByName());
    return axios(opts)
      .then(function(response) {
        let data = [];

        if (response.data) {
          data = response.data.results.map(createCharacterData);

          dispatch(receiveCharacters(data));
        }
      })
      .catch(function(response){
        dispatch(receiveError(response.data));
      });
  };
}
