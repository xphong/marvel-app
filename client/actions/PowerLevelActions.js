import axios from 'axios';

import * as types from '../constants/ActionTypes';
import { ENDPOINT } from '../constants/AppConstants';

function requestPowerLevels() {
  return {
    type: types.REQUEST_POWERLEVELS
  };
}

function receivePowerLevels(data) {
  return {
    type: types.RECEIVE_POWERLEVELS,
    payload: {
      data
    }
  };
}

function receivePowerLevelsError(data) {
  return {
    type: types.RECEIVE_POWERLEVELS_ERROR,
    payload: {
      data
    }
  };
}

function createPowerLevelData(powerlevel) {
  powerlevel.AveragePowerLevel = calculatePowerLevel(powerlevel);
  return powerlevel;
}

function calculatePowerLevel(character) {
  const numberOfSkills = 6;

  return ((parseInt(character.Intelligence)
    + parseInt(character.Strength)
    + parseInt(character.Speed)
    + parseInt(character.Durability)
    + parseInt(character.Energy_Projection)
    + parseInt(character.Fighting_Ability)) / numberOfSkills).toFixed(2);
}

function sortCharactersByName(data) {
  var sortedData = data.slice(0);

  sortedData.sort((a, b) => {
    var textA = a.Name.toUpperCase();
    var textB = b.Name.toUpperCase();
    return (textA < textB) ? -1 : (textA > textB) ? 1 : 0;
  });

  return sortedData;
}

export function fetchPowerLevels() {
  const url = `${ENDPOINT}/powerlevels`;
  const opts = {
    url: url,
    timeout: 10000,
    method: 'get',
    responseType: 'json'
  };

  return dispatch => {
    const powerLevelsSessionData = sessionStorage.getItem('powerlevels-data');

    dispatch(requestPowerLevels());

    if (powerLevelsSessionData) {
      return dispatch(receivePowerLevels(JSON.parse(powerLevelsSessionData)));
    }

    return axios(opts)
      .then(function(response) {
        let powerLevelsData = [];

        if (response.data && response.data.length) {
          powerLevelsData = response.data;
          powerLevelsData.map(createPowerLevelData);
          powerLevelsData = sortCharactersByName(response.data);

          sessionStorage.setItem('powerlevels-data', JSON.stringify(powerLevelsData));

          dispatch(receivePowerLevels(powerLevelsData));
        } else {
          dispatch(receivePowerLevelsError(response.data));
        }
      })
      .catch(function(response){
        dispatch(receivePowerLevelsError(response.data));
      });
  };
}
