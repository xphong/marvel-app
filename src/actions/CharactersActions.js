import axios from 'axios';

import * as types from '../constants/ActionTypes';
import { ENDPOINT, KEY, CHARACTER_LIMIT } from '../constants/AppConstants';

function requestCharactersByName() {
  return {
    type: types.REQUEST_CHARACTERS_BY_NAME
  };
}

function receiveCharacters(data) {
  return{
    type: types.RECEIVE_CHARACTERS,
    payload: {
      data
    }
  };
}

function receiveError(data) {
  return {
    type: types.RECEIVE_CHARACTERS_ERROR,
    payload: {
      data
    }
  };
}

function createCharacterData(data) {
  let characterData = {};

  characterData.name = data.name;
  characterData.url = (data.urls[1] ? data.urls[1].url : data.urls[0].url);
  characterData.image = data.thumbnail.path + '.' + data.thumbnail.extension;
  characterData.description =  (data.description === '' ? 'No description listed for this character.' : characterData.description = data.description);

  return characterData;
}

export function fetchCharactersByName(name) {
  const timestamp = new Date().getTime();
  const url = `${ENDPOINT}characters?limit=${CHARACTER_LIMIT}&nameStartsWith=${name}&apikey=${KEY}&ts=${timestamp}`;

  return function(dispatch) {
    dispatch(requestCharactersByName());
    return axios({
      url: url,
      timeout: 10000,
      method: 'get',
      responseType: 'json'
    })
    .then(function(response) {
      let data = [];

      if (response.data) {
        data = response.data.data.results.map(createCharacterData);

        dispatch(receiveCharacters(data));
      }
    })
    .catch(function(response){
      dispatch(receiveError(response.data));
    });
  };
}
