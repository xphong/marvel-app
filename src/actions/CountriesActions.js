import axios from 'axios';

import * as types from '../constants/ActionTypes';

function requestCountries() {
  return {
    type: types.REQUEST_COUNTRIES
  };
}

function receiveCountries(data) {
  return{
    type: types.RECEIVE_COUNTRIES,
    payload: {
      data
    }
  };
}

function receiveError(data) {
  return {
    type: types.RECEIVE_ERROR,
    payload: {
      data
    }
  };
}

export function fetchCountries(url) {
  return function(dispatch) {
    dispatch(requestCountries());
    return axios({
      url: url,
      timeout: 20000,
      method: 'get',
      responseType: 'json'
    })
    .then(function(response) {
      dispatch(receiveCountries(response.data));
    })
    .catch(function(response){
      dispatch(receiveError(response.data));
    });
  };
}
