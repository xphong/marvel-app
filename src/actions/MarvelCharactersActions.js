import axios from 'axios';

import * as types from '../constants/ActionTypes';
import { ENDPOINT, KEY, CHARACTER_LIMIT } from '../constants/AppConstants';

function requestMarvelCharactersByName() {
  return {
    type: types.REQUEST_MARVEL_CHARACTERS_BY_NAME
  };
}

function receiveMarvelCharacters(data) {
  return{
    type: types.RECEIVE_MARVEL_CHARACTERS,
    payload: {
      data
    }
  };
}

function receiveError(data) {
  return {
    type: types.RECEIVE_MARVEL_CHARACTERS_ERROR,
    payload: {
      data
    }
  };
}

export function fetchMarvelCharactersByName(name) {
  const timestamp = new Date().getTime();
  const url = `${ENDPOINT}characters?limit=${CHARACTER_LIMIT}&nameStartsWith=${name}&apikey=${KEY}&ts=${timestamp}`;

  return function(dispatch) {
    dispatch(requestMarvelCharactersByName());
    return axios({
      url: url,
      timeout: 10000,
      method: 'get',
      responseType: 'json'
    })
    .then(function(response) {
      dispatch(receiveMarvelCharacters(response.data.data.results));
    })
    .catch(function(response){
      dispatch(receiveError(response.data));
    });
  };
}
