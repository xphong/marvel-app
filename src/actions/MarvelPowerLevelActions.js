import axios from 'axios';

import * as types from '../constants/ActionTypes';
import { POWER_LEVELS } from '../constants/AppDataPowerLevels';

function requestPowerLevels() {
  return {
    type: types.REQUEST_MARVEL_POWERLEVELS
  };
}

function receivePowerLevels(data) {
  return{
    type: types.RECEIVE_MARVEL_POWERLEVELS,
    payload: {
      data
    }
  };
}

function receivePowerLevelsError(data) {
  return {
    type: types.RECEIVE_MARVEL_POWERLEVELS_ERROR,
    payload: {
      data
    }
  };
}

export function fetchMarvelPowerLevels() {
  return function(dispatch) {
    dispatch(receivePowerLevels(POWER_LEVELS));
  };
}
