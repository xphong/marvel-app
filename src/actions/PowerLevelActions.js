import axios from 'axios';

import * as types from '../constants/ActionTypes';
import { POWER_LEVELS } from '../constants/AppDataPowerLevels';

function requestPowerLevels() {
  return {
    type: types.REQUEST_POWERLEVELS
  };
}

function receivePowerLevels(data) {
  return {
    type: types.RECEIVE_POWERLEVELS,
    payload: {
      data
    }
  };
}

function receivePowerLevelsError(data) {
  return {
    type: types.RECEIVE_POWERLEVELS_ERROR,
    payload: {
      data
    }
  };
}

function createPowerLevelData(powerLevelKey) {
  let powerLevelData = POWER_LEVELS[powerLevelKey];
  powerLevelData.Name = powerLevelKey;
  powerLevelData.AveragePowerLevel = calculatePowerLevel(powerLevelData);
  return powerLevelData;
}

function calculatePowerLevel(character) {
  const numberOfSkills = 6;

  return ((parseInt(character.Intelligence)
    + parseInt(character.Strength)
    + parseInt(character.Speed)
    + parseInt(character.Durability)
    + parseInt(character.Energy_Projection)
    + parseInt(character.Fighting_Ability)) / numberOfSkills).toFixed(2);
}

function sortCharactersByName(data) {
  var sortedData = data.slice(0);

  sortedData.sort((a, b) => {
    var textA = a.Name.toUpperCase();
    var textB = b.Name.toUpperCase();
    return (textA < textB) ? -1 : (textA > textB) ? 1 : 0;
  });

  return sortedData;
}

export function fetchPowerLevels() {
  return function(dispatch) {
    let powerLevelsData = Object.keys(POWER_LEVELS).map(createPowerLevelData);

    powerLevelsData = sortCharactersByName(powerLevelsData);

    dispatch(receivePowerLevels(powerLevelsData));
  };
}
