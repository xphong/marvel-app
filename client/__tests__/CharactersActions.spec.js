import configureMockStore from 'redux-mock-store';
import moxios from 'moxios';
import thunk from 'redux-thunk';

import * as CharactersActions from '../actions/CharactersActions';
import * as mocks from './mocks/charactersMock';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Characters Actions: fetchCharactersByName', () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('should get characters by name', () => {
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: mocks.charactersAPIMock
      });
    });

    const store = mockStore({});

    return store.dispatch(CharactersActions.fetchCharactersByName('Venom'))
    .then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
