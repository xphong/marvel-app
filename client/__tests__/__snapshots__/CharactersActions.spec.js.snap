// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters Actions: fetchCharactersByName should get characters by name 1`] = `
Array [
  Object {
    "type": "@@marvel/REQUEST_CHARACTERS_BY_NAME",
  },
  Object {
    "payload": Object {
      "data": Array [
        Object {
          "description": "Flash Thompson was the high school bully of Peter Parker",
          "image": "http://i.annihil.us/u/prod/marvel/i/mg/7/80/5269668c7d5df.jpg",
          "name": "Venom",
          "url": "http://marvel.com/comics/characters/1009663/venom_flash_thompson",
        },
      ],
    },
    "type": "@@marvel/RECEIVE_CHARACTERS",
  },
]
`;
